### PR TITLE
TardisManager performance

### PR DESCRIPTION
**Pull request by DrTheodor**
Please test if you find anything bad.

### CODE REVIEW

File: src/main/java/mdteam/ait/core/blockentities/door/AbstractDoorBlockEntity.java [hunk 1]
```diff
@@ -54,7 +54,8 @@ public void onEntityCollision(Entity entity) {
     protected abstract void teleport(Entity entity);
 
     @Override
-    public void setTardis(Tardis tardis) {
+    public void setTardis(ITardis tardis) {
+        System.out.println("Set tardis to " + tardis);
```
Comment (Suff99) [inline src/main/java/mdteam/ait/core/blockentities/door/AbstractDoorBlockEntity.java:58]: Use a proper logger if this is going into production

File: src/main/java/mdteam/ait/tardis/TardisDesktop.java [hunk 2]
```diff
@@ -15,13 +15,17 @@ public class TardisDesktop extends AbstractTardisComponent {
...
+        super(tardis, "desktop");
+
+        this.schema = schema;
+        this.corners = TardisUtil.findInteriorSpot();
     }
 
     @Override
     public void init() {
-        BlockPos doorPos = new DesktopGenerator(schema).place(
+        System.out.println("Init desktop!");
```
Comment (Suff99) [inline src/main/java/mdteam/ait/tardis/TardisDesktop.java:27]: Logger!

**Review by Suff99 — COMMENTED** [review]
This is my worst nightmare

File: src/main/java/mdteam/ait/tardis/wrapper/server/manager/ServerTardisManager.java [hunk 3]
```diff
@@ -82,9 +77,11 @@ private Tardis loadTardis(UUID uuid) {
                 throw new IOException("Tardis file " + file + " doesn't exist!");
 
             String json = Files.readString(file.toPath());
-            Tardis tardis = this.gson.fromJson(json, Tardis.class);
-            this.lookup.put(tardis.getUuid(), tardis);
 
+            ServerTardis tardis = this.gson.fromJson(json, ServerTardis.class);
```
[the remark this answers is not in the document]
Comment (Suff99) [reply src/main/java/mdteam/ait/tardis/wrapper/server/manager/ServerTardisManager.java:81]: we're not in the dark ages anymore

[the remark this answers is not in the document]
Comment (Duzos) [reply src/main/java/mdteam/ait/tardis/wrapper/server/manager/ServerTardisManager.java:81]: nahh
saving with nbt is annoying

[the remark this answers is not in the document]
Comment (DrTheodor) [reply src/main/java/mdteam/ait/tardis/wrapper/server/manager/ServerTardisManager.java:81]: I told Loqor that using json to save stuff is bad. I proposed nbt but he declined saying that gson works well enough so

File: src/main/java/mdteam/ait/core/blockentities/door/AbstractDoorBlockEntity.java (hunk 1 above)
Comment (DrTheodor) [reply src/main/java/mdteam/ait/core/blockentities/door/AbstractDoorBlockEntity.java:58]: Yeah no this is WIP

File: src/main/java/mdteam/ait/tardis/AbstractTardisComponent.java [hunk 4]
```diff
@@ -9,25 +9,36 @@
 public abstract class AbstractTardisComponent {
 
     @Exclude
-    protected final Tardis tardis;
-    private final boolean shouldInit;
+    private ITardis tardis;
+    private final String id;
```
[the remark this answers is not in the document]
Comment (DrTheodor) [reply src/main/java/mdteam/ait/tardis/AbstractTardisComponent.java:13]: For the id field? I probably should've named it "name" to make it more obvious.

File: src/main/java/mdteam/ait/tardis/TardisTravel.java [hunk 5]
```diff
@@ -158,7 +157,7 @@ public void placeExterior() {
         );
 
         // this is needed for the initialization. when we call #setTardis(ITardis) the travel field is still null.
-        exterior.setTardis(this.tardis);
+        exterior.setTardis(this.getTardis());
         this.position.addBlockEntity(exterior);
```
[the remark this answers is not in the document]
Comment (DrTheodor) [reply src/main/java/mdteam/ait/tardis/TardisTravel.java:161]: It's actually totally fine, LinkableBlockEntity and TardisManagers can handle this.

File: src/main/java/mdteam/ait/tardis/wrapper/client/ClientTardis.java [hunk 6]
```diff
@@ -5,10 +5,70 @@
...
+    private TardisDesktop desktop;
+    private TardisExterior exterior;
+    private TardisDoor door;
 
     public ClientTardis(UUID uuid, AbsoluteBlockPos.Directed pos, TardisDesktopSchema schema, ExteriorEnum exteriorType) {
-        super(uuid, tardis -> new TardisTravel(tardis, pos), tardis -> new TardisDesktop(tardis, schema), tardis -> new ClientTardisExterior(tardis, exteriorType), TardisDoor::new);
+        this(uuid, tardis -> new TardisTravel(tardis, pos), tardis -> new TardisDesktop(tardis, schema), tardis -> new TardisExterior(tardis, exteriorType), TardisDoor::new);
+    }
+
+    protected ClientTardis(UUID uuid, Function<ITardis, TardisTravel> travel, Function<ITardis, TardisDesktop> desktop,
```
[the remark this answers is not in the document]
Comment (DrTheodor) [reply src/main/java/mdteam/ait/tardis/wrapper/client/ClientTardis.java:22]: It was a leftover of the previous tardis code that was split into client and server, gonna remove the unneeded constructors later.

File: src/main/java/mdteam/ait/tardis/wrapper/server/ServerTardis.java [hunk 7]
```diff
@@ -2,24 +2,66 @@
...
+
+    @Override
+    public TardisDesktop getDesktop() {
+        return desktop;
+    }
+
     @Override
     public void setDesktop(TardisDesktop desktop) {
-        super.setDesktop(desktop);
+        this.desktop = (ServerTardisDesktop) desktop;
```
[the remark this answers is not in the document]
Comment (DrTheodor) [reply src/main/java/mdteam/ait/tardis/wrapper/server/ServerTardis.java:64]: Fair

**Comment by Duzos** [comment]
oops what does that do

File: src/main/java/mdteam/ait/core/blockentities/door/AbstractDoorBlockEntity.java (hunk 1 above)
Comment (Duzos) [reply src/main/java/mdteam/ait/core/blockentities/door/AbstractDoorBlockEntity.java:58]: WOAH COMMENT FEATRURE HOW COOL!!

File: src/main/java/mdteam/ait/tardis/wrapper/server/manager/ServerTardisManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Duzos) [reply src/main/java/mdteam/ait/tardis/wrapper/server/manager/ServerTardisManager.java:81]: we will be moving to nbt

**Review by Duzos — APPROVED** [review]
code wise everything is ok except what craig mentioned
lots of bugs tho
like tp bugs are still a thing on this branch
tardis doors and exteriors going invisible
loqors still testing so hes writing his stuff now
i think we merge this into main and then we merge duzo/branch into main afterwards and deal with that chaos
basically im just hopeful things from my branch will fix the large issues in this one
i approve

**Review by Loqor — CHANGES_REQUESTED** [review]
The main issue is reliability, and a few bugs here are fixed on duzos/branch as of today. Linkable.getTardis() is null after relog, then linking the remote and using it causes the timer to error and this.tardis to be null, and once the tardis is landed, it crashes the game immediately. The models of the exterior and interior doors also go completely invisible, and the doors have weird double opening issues as well. Overall, this needs to be merged with the current changes first then worked on afterwards. Otherwise, my only issues are that it hasn't been worked on since last week.

**Pull request closed without merging**